flashinfer/gemm.py

```diff
@@ -628,13 +628,158 @@ def run(
+            f"Only torch.bfloat16 and torch.float16 are supported for FP8 GEMM operations."
+        )
+
+
+@functools.lru_cache(maxsize=128)
+def build_cudnn_gemm_with_per_tensor_q_graph(
+    a_shape, a_stride, b_shape, b_stride, a_type, b_type, o_type
+):
+    """Build a cuDNN graph for GEMM with per-tensor quantization.
+
+    This function is cached to avoid rebuilding identical graphs.
+
+    Args:
+        a_shape: Shape of tensor A
+        a_stride: Stride of tensor A
+        b_shape: Shape of tensor B
+        b_stride: Stride of tensor B
+        a_type: Data type for input tensor A
+        b_type: Data type for input tensor B
+        o_type: Data type for output tensor
+
+    Returns:
+        cuDNN graph object
+    """
+    _check_cudnn_availability()
+
+    graph = cudnn.pygraph()
+
+    a_cudnn_tensor = graph.tensor(
+        name="a", dim=a_shape, stride=a_stride, data_type=a_type
+    )
+    b_cudnn_tensor = graph.tensor(
+        name="b", dim=b_shape, stride=b_stride, data_type=b_type
+    )
+    scale_cudnn_tensor = graph.tensor(
+        name="scale", dim=(1, 1, 1), stride=(1, 1, 1), data_type=cudnn.data_type.FLOAT
+    )
+    c_cudnn_tensor = graph.matmul(
+        name="matmul",
+        A=a_cudnn_tensor,
+        B=b_cudnn_tensor,
+        compute_data_type=cudnn.data_type.FLOAT,
+    )
+    c_cudnn_tensor.set_name("c").set_data_type(cudnn.data_type.FLOAT)
+    c_final_cudnn_tensor = graph.mul(
+        name="scale_mul",
+        a=c_cudnn_tensor,
+        b=scale_cudnn_tensor,
+        compute_data_type=cudnn.data_type.FLOAT,
+    )
+    c_final_cudnn_tensor.set_name("c_final").set_output(True).set_data_type(o_type)
+
+    a_cudnn_tensor.set_uid(UIDs.A_UID.value)
+    b_cudnn_tensor.set_uid(UIDs.B_UID.value)
+    scale_cudnn_tensor.set_uid(UIDs.SCALE_UID.value)
+    c_final_cudnn_tensor.set_uid(UIDs.O_UID.value)
+
+    graph.validate()
+    graph.build_operation_graph()
+    graph.create_execution_plans([cudnn.heur_mode.A, cudnn.heur_mode.FALLBACK])
+    graph.check_support()
+    graph.build_plans()
+
+    return graph
+
+
+def execute_cudnn_gemm_with_per_tensor_q_graph(graph, a, b, scale_tensor, c_final):
+    variant_pack = {
+        UIDs.A_UID.value: a,
+        UIDs.B_UID.value: b,
+        UIDs.SCALE_UID.value: scale_tensor,
+        UIDs.O_UID.value: c_final,
+    }
+
+    workspace = torch.empty(
+        graph.get_workspace_size(), device="cuda", dtype=torch.uint8
+    )
+
+    graph.execute(variant_pack, workspace)
+
+
+def _torch_data_type_to_cudnn_data_type(dtype: torch.dtype):
+    if dtype == torch.bfloat16:
+        return cudnn.data_type.BFLOAT16
+    elif dtype == torch.float16:
+        return cudnn.data_type.HALF
+    elif dtype == torch.float8_e4m3fn:
+        return cudnn.data_type.FP8_E4M3
+    elif dtype == torch.float8_e5m2:
+        return cudnn.data_type.FP8_E5M2
+    else:
+        raise ValueError(f"Unsupported dtype: {dtype}")
+
+
+def _cudnn_gemm_f8f8_f32(
```

Comment:
here the f32 means the compute dtype, not output, so it might be confusing, we can either indicate it's compute dtype, or safely ignore it (most of our kernels use f32 as compute dtypes internally).